middleware/security/jwt/jwt.go

```diff
@@ -2,16 +2,153 @@ package jwt
+// *rsa.PublicKey or []*rsa.PublicKey.
+var ErrInvalidKey = errors.New("invalid parameter, the only keys accepted " +
+	"are *rsa.publicKey, []*rsa.PublicKey (for RSA-based algorithms) or a " +
+	"signing secret string, []string (for HS algorithms)")
+
+// ErrKeyDoesNotExist is returned when a key cannot be found by the provided
+// key name.
+var ErrKeyDoesNotExist = errors.New("key does not exist")
+
+// KeyResolver is a struct that is passed into the New() function, which allows
+// the user to add/remove keys from the jwt goa.middleware. The use of a
+// resolver provides for better scalability/performance as the number of
+// valid keys grows. If an incoming http.Request contains the header field
+// "jwtkeyname", then the handler will only attempt to validate the incoming
+// JWT against the keys stored in the resolver under that name. Otherwise,
+// the handler will attempt to validate the incoming JWT against all keys
+// stored in the resolver.
+type KeyResolver struct {
+	*sync.Mutex
+	keyMap map[string][]interface{}
+}
+
+// AddKeys can be used to add keys to the resolver which will be referenced
+// by the provided name. Acceptable types for keys include string, []string,
+// *rsa.PublicKey or []*rsa.PublicKey. Multiple keys are allowed for a single
+// key name to allow for key rotation.
+func (kr *KeyResolver) AddKeys(name string, keys interface{}) error {
+	kr.Lock()
+	defer kr.Unlock()
+	if _, ok := kr.keyMap[name]; !ok {
+		kr.keyMap[name] = make([]interface{}, 0)
+	}
+
+	switch keys := keys.(type) {
+	case *rsa.PublicKey:
+		kr.keyMap[name] = append(kr.keyMap[name], keys)
+	case []*rsa.PublicKey:
+		for _, key := range keys {
+			kr.keyMap[name] = append(kr.keyMap[name], key)
+		}
+	case string:
+		kr.keyMap[name] = append(kr.keyMap[name], keys)
+	case []string:
+		for _, key := range keys {
+			kr.keyMap[name] = append(kr.keyMap[name], key)
+		}
+	default:
+		return ErrInvalidKey
+	}
+	return nil
+}
+
+// RemoveAllKeys removes all keys from the resolver.
+func (kr *KeyResolver) RemoveAllKeys() {
+	kr.Lock()
+	defer kr.Unlock()
+	kr.keyMap = make(map[string][]interface{})
+	return
+}
+
+// RemoveKeys removes all keys from the resolver stored under the provided name.
+func (kr *KeyResolver) RemoveKeys(name string) {
+	kr.Lock()
+	defer kr.Unlock()
+	delete(kr.keyMap, name)
+	return
+}
+
+// RemoveKey removes only the provided key stored under the provided name from
+// the resolver.
+func (kr *KeyResolver) RemoveKey(name string, key interface{}) {
+	kr.Lock()
+	defer kr.Unlock()
+	if keys, ok := kr.keyMap[name]; ok {
+		for i, keyItem := range keys {
+			if keyItem == key {
+				kr.keyMap[name] = append(keys[:i], keys[i+1:]...)
+			}
+		}
+	}
+	return
+}
+
+// GetAllKeys returns a list of all the keys stored in the resolver.
+func (kr *KeyResolver) GetAllKeys() []interface{} {
+	kr.Lock()
+	defer kr.Unlock()
+	var keys []interface{}
+	for name := range kr.keyMap {
```

Comment:
Same as above, my suggestion doesn't work because the slice elements have different types...